File: apps/backend-python/src/routes/api/v1/tags.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from typing import Optional
from sqlalchemy.orm import Session
from src.config.database import get_db
from src.middleware.auth import get_current_user
from src.services.tag_service import tag_service
from src.db.models import User

router = APIRouter()


class CreateTagSchema(BaseModel):
    name: str
    data: Optional[dict] = None


@router.post("")
def create_tag(data: CreateTagSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return tag_service.create_tag(db, user.id, data.name, data.data)


@router.get("")
def list_tags(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return tag_service.get_tags(db, user.id)


@router.get("/{tag_id}")
def get_tag(tag_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return tag_service.get_tag_by_id(db, tag_id, user.id)


@router.delete("/{tag_id}")
def delete_tag(tag_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    tag_service.delete_tag(db, tag_id, user.id)
    return {"message": "Tag deleted"}

File: apps/backend/src/routes/api/v1/tags.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { tagService } from '../../../services/tag.service';
import { authMiddleware } from '../../../middleware/auth';
import { createTagSchema } from '../../../utils/validators-extended';

const tagRouter = new Hono();

tagRouter.post('/', authMiddleware, zValidator('json', createTagSchema), async (c) => {
  const user = c.get('user');
  const { name, data } = c.req.valid('json');

  const tag = await tagService.createTag(user.id, name, data);
  return c.json(tag);
});

tagRouter.get('/', authMiddleware, async (c) => {
  const user = c.get('user');
  const tags = await tagService.getTags(user.id);
  return c.json(tags);
});

tagRouter.get('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  const tag = await tagService.getTagById(id, user.id);
  return c.json(tag);
});

tagRouter.delete('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  await tagService.deleteTag(id, user.id);
  return c.json({ message: 'Tag deleted' });
});

export default tagRouter;
